pipeline-runner/R/seurat-2-load_seurat.R

```diff
@@ -53,53 +53,75 @@ reconstruct_seurat <- function(dataset_fpath) {
 
   # get counts
   tryCatch({
+    SeuratObject::DefaultAssay(user_scdata) <- 'RNA'
     counts <- user_scdata[['RNA']]@counts
     test_user_sparse_mat(counts)
     rns <- row.names(counts)
     check_type_is_safe(rns)
-  },
-  error = function(e) {
-    message(e$message)
-    stop(errors$ERROR_SEURAT_COUNTS, call. = FALSE)
-  })
-
-  gene_annotations <- data.frame(input = rns, name = rns, original_name = rns, row.names = rns)
-  user_scdata@misc$gene_annotations <- gene_annotations
 
-  # add dispersions
-  tryCatch({
-    user_scdata <- add_dispersions(user_scdata, method = "default")
-    dispersions <- user_scdata@misc$gene_dispersion
-    test_user_df(dispersions)
   },
   error = function(e) {
     message(e$message)
-    stop(errors$ERROR_SEURAT_HVFINFO, call. = FALSE)
+    stop(errors$ERROR_SEURAT_COUNTS, call. = FALSE)
   })
 
   # get meta data
   tryCatch({
     metadata <- user_scdata@meta.data
+    metadata$active.ident <- user_scdata@active.ident
     test_user_df(metadata)
   },
   error = function(e) {
     message(e$message)
     stop(errors$ERROR_SEURAT_METADATA, call. = FALSE)
   })
 
-  # check for clusters
-  if (!'seurat_clusters' %in% colnames(metadata))
-    stop(errors$ERROR_SEURAT_CLUSTERS, call. = FALSE)
-
-
-  # reconstruct seurat object
+  # reconstruct Seurat object
   scdata <- SeuratObject::CreateSeuratObject(
     counts,
     meta.data = metadata,
   )
 
-  # add dispersions and gene annotations to new scdata
-  scdata@misc$gene_dispersion <- dispersions
+  # add logcounts
+  tryCatch({
+    logcounts <- user_scdata[['RNA']]@data
+    test_user_sparse_mat(logcounts)
+
+    # shouldn't be raw counts
+    suspect.counts <- max(logcounts) > 100
+    if (suspect.counts) logcounts <- Seurat::NormalizeData(logcounts)
+
+    scdata[['RNA']]@data <- logcounts
+  },
+  error = function(e) {
+    message(e$message)
+    stop(errors$ERROR_SEURAT_LOGCOUNTS, call. = FALSE)
+  })
+
+
+  # add dispersions (need for gene list)
+  tryCatch({
+    # don't use first 3 columns
+    dispersions <- Seurat::FindVariableFeatures(logcounts)[-3]
+
+    test_user_df(dispersions)
+    colnames(dispersions) <- gsub('^vst[.]', '', colnames(dispersions))
+    dispersions$SYMBOL <- dispersions$ENSEMBL <- row.names(dispersions)
+
+    scdata@misc$gene_dispersion <- dispersions
+  },
+  error = function(e) {
+    message(e$message)
+    stop(errors$ERROR_SEURAT_HVFINFO, call. = FALSE)
+  })
+
+  # add gene annotations
+  gene_annotations <- data.frame(
+    input = rns,
+    name = rns,
+    original_name = rns,
+    row.names = rns
```

Comment:
you could check if the gene_annotations exist in the seurat object, and just use that (testing for df safety). Or are gene names handled elsewhere?

This would add support for rds objects downloaded from Cellenics (some extra step might be required downstream though, don't remember now).

Reply:
I don't really see that as the use-case of Seurat object upload. If the project was processed in Cellenics, you should look at the original project so that you can have access to all of the processing options and associated information in Data Processing.

Reply:
given this, I'd say it's good to go!

Reply:
I added it -- was relatively simple and I think it makes sense from the perspective that it simplifies things for developers (can use Seurat objects downloaded from Cellenics for testing). Thanks for the suggestion!